src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -305,6 +305,59 @@ private async Task ReplyOnReAuthenticationAsync<TIOAdapter>(TIOAdapter adapter,
             }
         }
 
+        // This will initiate renegotiation or PHA for Tls1.3
+        private async Task RenegotiateAsync(CancellationToken cancellationToken)
+        {
+            if (Interlocked.Exchange(ref _nestedAuth, 1) == 1)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidnestedcall, "NegotiateClientCertificateAsync", "renegotiate"));
+            }
+
+            if (Interlocked.CompareExchange(ref _nestedRead, 1, 0) == 1)
```

Comment:
Are there values _nestedRead can be other than 0 and 1?  I'm wondering why this is using CompareExchange whereas the check above for _nestedAuth is used Exchange.

Reply:
the values are only 0 and 1 and there is no real reason why the operations need to be different @stephentoub. The _nestedAuth is originally used during handshake and here it primarily serves as guard to allow  RenegotiateAsync only once. (we will not restore it to 0 once RenegotiateAsync is finished)
The _nestedRead and _nestedWrite guard agains race with read/write operations. 

I can update it for consistency. It seems lie we use Interlocked.CompareExchange in CloseInternal().

Reply:
updated.